'New Tree' Refactors

rburema opened the pull request:
Propose to keep the code bases closer together, so it will be easier to merge (new) fixes.

BagelOrb committed:
take disallowed areas into account when printing tree support

BagelOrb committed:
feat: settings.get<Polygons>

Piezoid committed:
Simplify: fix scaling of maximum area deviation setting

meshfix_maximum_extrusion_area_deviation is already in millimeters.
using get<coord_t> divides the value by 1000, instead it should be
get<size_t> or get<int>.

jellespijker committed:
Merge pull request #1738 from Piezoid/area_deviation_scaling

Simplify: fix scaling of maximum area deviation setting

rburema committed:
Revert "Merge pull request #1738 from Piezoid/area_deviation_scaling"

This reverts commit 1adf212782ff98bbf71dbb432e4ba0fe51ab468c, reversing
changes made to 1363b817eed41108acac22b760e338292678889d.

rburema committed:
Merge branch '5.2' into xmas_alpha

rburema committed:
Revert "Merge remote-tracking branch 'upstream/main' into tree_support_2"

This reverts commit e535e04f78585c7fb150f5d48b87d978c5d1e32d, reversing
changes made to 034fe146bcd3c44bc0f061be006044b3da4f3ded.

rburema committed:
New Tree Support: Quick refactors (except the big class).

part of CURA-9879

jellespijker committed:
Sync workflows from main

Contributes to CURA-9879

jellespijker committed:
Use 5.3.0-alpha dependencies

Revert before merging to main

Contributes to CURA-9879

jellespijker committed:
Use latest Arcus

Revert before merging to main

Contributes to CURA-9879

jellespijker committed:
Use PUBLIC targets for deps in _CuraEngine

Contributes to CURA-9879

jellespijker committed:
Use range-v3 instead of std

Mac is being Mac again, also the
range-v3 library is used consistently
throughout the rest of the code

Contributes to CURA-9879

rburema committed:
New Tree Support: Refactors part II.

Put code in our code style -- including moving comments around, for better readability so definitely not all that is said there is mine ;-)

part of CURA-9879

rburema committed:
New Tree Support: Refactors part III.

Put code in our code style -- including moving comments around, for better readability so definitely not all that is said there is mine ;-) -- Includes workaround for Mac compiler bug stumbled into in last commit.

part of CURA-9879

rburema committed:
Switch back to the original 'pass-by-copy' design.

Pass sby reference was erronously introduced by refactoring.

part of CURA-9879

ThomasRahm committed:
Fix xy_distance overriding overhangs when the overhang angle is smallish.

Manually cherry picked for CURA-9879 (git gave a bit of a mess when git cherry-pick was run...)

rburema committed:
Small refactor: Break some more comments up into lines.

part of CURA-9879

rburema committed:
Small refactors (mostly re-lined comments).

part of CURA-9879

rburema committed:
Forgot to rename this instance of mesh for the apple-clang workaround.

part of CURA-9879

rburema committed:
Refactor: Use 'using' to abreviate often used types.

part of CURA-9879

rburema committed:
Minimize use of (implict) sqrt operations.

Will not do that much, but standard in the rest of the code.

part of CURA-9879

rburema committed:
Replaced some of the more common magic numbers with constexprs.

part of CURA-9879

rburema committed:
Resolve temporary comments made during review.

part of CURA-9879

rburema committed:
'Workaround' for crash with tree-branch-pref set to 'flat surfaces'.

Added a note as well. I have no idea why the (in context) equivalent (even when the static_cast<size_t> is replaced by a simple cast to int) statement, using iota, doesn't work, but this does. It may just expose another bug, but best not to take the risk right now.

part of CURA-9879

rburema committed:
Undo small part of refactor that was causing issues.

It (moving the area-calculation to outside of the loop) _shoulnd't_ have caused issues (indicating that there's probably some deadlock possibility in the code as-is), but it _did_, so undo this part (for now).

part of CURA-9879

ThomasRahm commented:
This would also undo ["take disallowed areas into account when printing tree support"](https://github.com/Ultimaker/CuraEngine/commit/fb97bc7cc692552a6f6a2a96775524c2122502d8). Is this intentional or should i cherry pick this commit?

ThomasRahm commented on src/TreeSupport.cpp:
```diff
@@ -2246,42 +2504,46 @@ void TreeSupport::generateBranchAreas(std::vector<std::pair<LayerIndex, TreeSupp
-    // smooth upwards
-    for (LayerIndex layer_idx = 0; layer_idx < LayerIndex(layer_tree_polygons.size()) - 1; layer_idx++)
+    // Smooth upward.
+    for (const auto layer_idx : ranges::views::iota(0UL, layer_tree_polygons.size() - 1UL))
```
I do not think it is very relevant as i don't think that this code can even be reached in the relevant edge-case, but theoretically if `layer_tree_polygons.size() == 0` then this would underflow, while it would not underflow in my version as LayerIndex is a signed type.

ThomasRahm commented:
Why is PropertyAreas and PropertyAreasUnordered in TreeSupportElements.h and not inside TreeSupport.h? 
There may be a good reason why, it's just not where i would have been looking for it.

rburema committed:
Fix potential edge case: All but one of empty list.

rburema committed:
Moved PropertyAreasX definitions to where they're used.

rburema commented:
Hi @ThomasRahm . Sorry it took a bit longer to get back.

- It's not intentional to undo that commit. I've cherry-picked those necessary to this branch.
- I just fixed the 'empty lists' exception case with a little bit of defensive programming.
- `PropertyAreas` and `PropertyAreas` where indeed only used in `TreeSupport.___`, so I moved those over.

I should also really mention that our team has decided that 5.3 is already very 'full' of features, and we've got bad experiences with putting everything in at once. As such, we'd like to take some more time on the New Tree Support as well. So it'll be in 5.4. (We try to release every two to three months.) In the mean-time, the alpha is still out, so people can continue with that for a while.
